cmd/disk-cache-backend.go

```diff
@@ -0,0 +1,565 @@
+func (c *diskCache) IsOnline() bool {
+	c.onlineMutex.RLock()
+	defer c.onlineMutex.RUnlock()
+	return c.online
+}
+
+// Stat returns ObjectInfo from disk cache
+func (c *diskCache) Stat(ctx context.Context, bucket, object string) (oi ObjectInfo, err error) {
+	cachePath := c.getCacheSHADir(bucket, object)
+	cLock := c.nsMutex.NewNSLock(cachePath, "")
+	if err := cLock.GetRLock(globalObjectTimeout); err != nil {
+		return oi, err
+	}
+	defer cLock.RUnlock()
+
+	oi, err = c.stat(ctx, bucket, object)
+	return
+}
+
+// lockless helper to Stat
+func (c *diskCache) stat(ctx context.Context, bucket, object string) (oi ObjectInfo, e error) {
+	cacheObjPath := c.getCacheSHADir(bucket, object)
+	oi, e = c.statCache(ctx, cacheObjPath)
+	if e != nil {
+		return
+	}
+	oi.Bucket = bucket
+	oi.Name = object
+	return
+}
+
+// statCache is a convenience function for purge() to get ObjectInfo for cached object
+func (c *diskCache) statCache(ctx context.Context, cacheObjPath string) (oi ObjectInfo, e error) {
+	// Stat the file to get file size.
+	fi, err := fsStatFile(ctx, pathJoin(cacheObjPath, cacheDataFile))
+	if err != nil {
+		return oi, err
+	}
+	metaPath := pathJoin(cacheObjPath, cacheMetaJSONFile)
+
+	f, err := os.Open(metaPath)
+	if err != nil {
+		return oi, err
+	}
+	defer f.Close()
+	meta := &cacheMeta{Version: cacheMetaVersion}
+	if err := jsonLoad(f, meta); err != nil {
+		return oi, err
+	}
+	return meta.ToObjectInfo("", "", fi), nil
+}
+
+// caches object metadata to disk cache
+func (c *diskCache) saveMetadata(ctx context.Context, bucket, object string, meta map[string]string, partInfo []ObjectPartInfo) error {
+	fileName := c.getCacheSHADir(bucket, object)
+	metaPath := pathJoin(fileName, cacheMetaJSONFile)
+
+	f, err := os.Create(metaPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	m := cacheMeta{Meta: meta, Version: cacheMetaVersion, Parts: partInfo}
+	jsonData, err := json.Marshal(m)
+	if err != nil {
+		return err
+	}
+	f.Write(jsonData)
+	return nil
+}
+
+func (c *diskCache) getCacheSHADir(bucket, object string) string {
+	return path.Join(c.dir, getSHA256Hash([]byte(path.Join(bucket, object))))
+}
+
+// Caches the object to disk
+func (c *diskCache) Put(ctx context.Context, bucket, object string, data io.Reader, size int64, opts ObjectOptions) error {
+	if c.diskUsageHigh() {
+		select {
+		case c.purgeChan <- struct{}{}:
+		default:
+		}
+		return errDiskFull
+	}
+	if !c.diskAvailable(size) {
+		return errDiskFull
+	}
+	cachePath := c.getCacheSHADir(bucket, object)
+	cLock := c.nsMutex.NewNSLock(cachePath, "")
+	if err := cLock.GetLock(globalObjectTimeout); err != nil {
+		return err
+	}
+	defer cLock.Unlock()
+	if err := os.MkdirAll(cachePath, 0777); err != nil {
+		return err
+	}
+	bufSize := int64(readSizeV1)
+	if size > 0 && bufSize > size {
+		bufSize = size
+	}
+	filePath := path.Join(cachePath, cacheDataFile)
+	buf := make([]byte, int(bufSize))
+	_, err := fsCreateFile(ctx, filePath, data, buf, size)
```

Comment:
fsCreateFile() uses some fsObjects specific code (like checkDiskFree, lock.Open etc)

you can just use os.Create() directly here and write to it

Reply:
done